Remove hex crate dependency

ashishajr opened the pull request:
# Description
- Create `encode` to convert bytes to hexadecimal string
- Modify all `encode` to use custom implementation from `runner::hex`
- Remove `hex` crate as a dependency

# Related Issue
#4011 

# Discussion
Hey @jcamiel, let me know if this implementation of encode works or I need to make any changes

I had originally written the function as
```rust
pub fn encode(bytes: &[u8]) -> String {
    bytes.iter().map(|byte| format!("{:02x}", byte)).collect()
}
```
But I learnt this would be really inefficient so I changed it to new version (similar to how the `hex` crate does it)

Here's how the `hex` crate does `encode`

```rust
pub fn encode<T: AsRef<[u8]>>(data: T) -> String {
    data.encode_hex()
}

pub trait ToHex {
    /// Encode the hex strict representing `self` into the result. Lower case
    /// letters are used (e.g. `f9b4ca`)
    fn encode_hex<T: iter::FromIterator<char>>(&self) -> T;

    /// Encode the hex strict representing `self` into the result. Upper case
    /// letters are used (e.g. `F9B4CA`)
    fn encode_hex_upper<T: iter::FromIterator<char>>(&self) -> T;
}

const HEX_CHARS_LOWER: &[u8; 16] = b"0123456789abcdef";
const HEX_CHARS_UPPER: &[u8; 16] = b"0123456789ABCDEF";

struct BytesToHexChars<'a> {
    inner: ::core::slice::Iter<'a, u8>,
    table: &'static [u8; 16],
    next: Option<char>,
}

impl<'a> BytesToHexChars<'a> {
    fn new(inner: &'a [u8], table: &'static [u8; 16]) -> BytesToHexChars<'a> {
        BytesToHexChars {
            inner: inner.iter(),
            table,
            next: None,
        }
    }
}

impl<'a> Iterator for BytesToHexChars<'a> {
    type Item = char;

    fn next(&mut self) -> Option<Self::Item> {
        match self.next.take() {
            Some(current) => Some(current),
            None => self.inner.next().map(|byte| {
                let current = self.table[(byte >> 4) as usize] as char;
                self.next = Some(self.table[(byte & 0x0F) as usize] as char);
                current
            }),
        }
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        let length = self.len();
        (length, Some(length))
    }
}

impl<'a> iter::ExactSizeIterator for BytesToHexChars<'a> {
    fn len(&self) -> usize {
        let mut length = self.inner.len() * 2;
        if self.next.is_some() {
            length += 1;
        }
        length
    }
}

#[inline]
fn encode_to_iter<T: iter::FromIterator<char>>(table: &'static [u8; 16], source: &[u8]) -> T {
    BytesToHexChars::new(source, table).collect()
}

impl<T: AsRef<[u8]>> ToHex for T {
    fn encode_hex<U: iter::FromIterator<char>>(&self) -> U {
        encode_to_iter(HEX_CHARS_LOWER, self.as_ref())
    }

    fn encode_hex_upper<U: iter::FromIterator<char>>(&self) -> U {
        encode_to_iter(HEX_CHARS_UPPER, self.as_ref())
    }
}
```

If this looks good I'll replace the already existing references to `encode` to ours

jcamiel commented:
Hi @ashishajr 

Ok, but we can simplify a bit:

- no need to support for UPPERCASE, we don't want it for the moment:
  - so no need for `fn encode_to_iter`
  - so no need for `impl<T: AsRef<[u8]>> ToHex for T`
  - we can hardcode `HEX_CHARS_LOWER` as there is only one table
- no need to use `pub trait ToHex`

Basically, only the iterator `BytesToHexChars`:

```rust
pub fn encode(bytes: &[u8]) -> String {
    BytesToHexChars::new(source).collect()
}

const HEX_CHARS_LOWER: &[u8; 16] = b"0123456789abcdef";

struct BytesToHexChars<'a> {
    inner: ::core::slice::Iter<'a, u8>,
    next: Option<char>,
}

impl<'a> BytesToHexChars<'a> {
    fn new(inner: &'a [u8]) -> BytesToHexChars<'a> {
        BytesToHexChars { inner: inner.iter() }
    }
}

impl<'a> Iterator for BytesToHexChars<'a> {
    type Item = char;

    fn next(&mut self) -> Option<Self::Item> {
        match self.next.take() {
            Some(current) => Some(current),
            None => self.inner.next().map(|byte| {
                let current = HEX_CHARS_LOWER[(byte >> 4) as usize] as char;
                self.next = Some(HEX_CHARS_LOWER[(byte & 0x0F) as usize] as char);
                current
            }),
        }
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        let length = self.len();
        (length, Some(length))
    }
}

impl<'a> iter::ExactSizeIterator for BytesToHexChars<'a> {
    fn len(&self) -> usize {
        let mut length = self.inner.len() * 2;
        if self.next.is_some() {
            length += 1;
        }
        length
    }
}
```

ashishajr commented:
Thanks for the feedback @jcamiel, I've changed the implementation, I've also used this new implementation in `toHex` because cargo was complaining that the code wasn't used anywhere. If the new `encode` function looks good I can change the other references and squash commits.

jcamiel reviewed (changes requested):
@ashishajr `encode` implementation looks good, just a tiny modification for importing the function

ashishajr commented on packages/hurl/src/runner/filter/to_hex.rs:
```diff
@@ -27,7 +26,7 @@ pub fn eval_to_hex(
 ) -> Result<Option<Value>, RunnerError> {
     match value {
-        Value::Bytes(value) => Ok(Some(Value::String(hex::encode(value)))),
+        Value::Bytes(value) => Ok(Some(Value::String(encode(value)))),
```
Got it, thanks! Changed it to `hex::encode`

ashishajr commented:
Thanks @jcamiel, I've updated all references to use `runner::encode` and removed the `hex` crate from `Cargo.lock` as well. Please do let me know if in case I missed something

jcamiel reviewed (approved):
Everything is good thanks @ashishajr 👍

jcamiel commented:
/accept
